<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 The AnVIL R / Bioconductor package | R / Bioconductor in the AnVIL Cloud</title>
  <meta name="description" content="This book contains material for a 2 1/2 hour course offered at the Bioinformatics Community Conference 2020." />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 The AnVIL R / Bioconductor package | R / Bioconductor in the AnVIL Cloud" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This book contains material for a 2 1/2 hour course offered at the Bioinformatics Community Conference 2020." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 The AnVIL R / Bioconductor package | R / Bioconductor in the AnVIL Cloud" />
  
  <meta name="twitter:description" content="This book contains material for a 2 1/2 hour course offered at the Bioinformatics Community Conference 2020." />
  

<meta name="author" content="Martin Morgan" />
<meta name="author" content="Nitesh Turaga" />
<meta name="author" content="Lori Shepherd" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="working-with-large-data.html"/>
<link rel="next" href="formal-work-flows-human-cell-atlas-optimus-pipeline.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">R / Bioconductor in the AnVIL Cloud</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a><ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i>Abstract</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i>Prerequisites</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#learning-objectives"><i class="fa fa-check"></i>Learning objectives</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1</b> Prerequisites</a><ul>
<li class="chapter" data-level="1.1" data-path="prerequisites.html"><a href="prerequisites.html"><i class="fa fa-check"></i><b>1.1</b> Create a google ‘billing account’</a></li>
<li class="chapter" data-level="1.2" data-path="prerequisites.html"><a href="prerequisites.html#create-an-anvil-account"><i class="fa fa-check"></i><b>1.2</b> Create an AnVIL account</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html"><i class="fa fa-check"></i><b>2</b> The AnVIL – workspaces, runtimes, buckets, workflows</a><ul>
<li class="chapter" data-level="2.1" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html#navigation-icons"><i class="fa fa-check"></i><b>2.1</b> Navigation icons</a></li>
<li class="chapter" data-level="2.2" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html#use-an-existing-workspace"><i class="fa fa-check"></i><b>2.2</b> Use an existing workspace</a></li>
<li class="chapter" data-level="2.3" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html#jupyter-notebook-runtimes"><i class="fa fa-check"></i><b>2.3</b> Jupyter notebook runtimes</a></li>
<li class="chapter" data-level="2.4" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html#google-bucket"><i class="fa fa-check"></i><b>2.4</b> Google bucket</a></li>
<li class="chapter" data-level="2.5" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html#notebooks"><i class="fa fa-check"></i><b>2.5</b> Notebooks</a></li>
<li class="chapter" data-level="2.6" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html#workflows"><i class="fa fa-check"></i><b>2.6</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html"><i class="fa fa-check"></i><b>3</b> Using an existing workspace</a><ul>
<li class="chapter" data-level="3.1" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#orchestrating-single-cell-analysis"><i class="fa fa-check"></i><b>3.1</b> Orchestrating single cell analysis</a></li>
<li class="chapter" data-level="3.2" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#install-and-load-r-packages"><i class="fa fa-check"></i><b>3.2</b> Install and load <em>R</em> packages</a></li>
<li class="chapter" data-level="3.3" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#example-data-as-a-singlecellexperiment"><i class="fa fa-check"></i><b>3.3</b> Example data as a SingleCellExperiment</a></li>
<li class="chapter" data-level="3.4" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#the-quick-start-workflow"><i class="fa fa-check"></i><b>3.4</b> The ‘Quick start’ workflow</a></li>
<li class="chapter" data-level="3.5" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#information-about-the-packages-used-in-this-session"><i class="fa fa-check"></i><b>3.5</b> Information about the packages used in this session</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="working-with-large-data.html"><a href="working-with-large-data.html"><i class="fa fa-check"></i><b>4</b> Working with large data</a><ul>
<li class="chapter" data-level="4.1" data-path="working-with-large-data.html"><a href="working-with-large-data.html#human-cell-atlas-data-discovery"><i class="fa fa-check"></i><b>4.1</b> Human cell atlas data discovery</a></li>
<li class="chapter" data-level="4.2" data-path="working-with-large-data.html"><a href="working-with-large-data.html#r-bioconductor-libraries"><i class="fa fa-check"></i><b>4.2</b> <em>R / Bioconductor</em> libraries</a></li>
<li class="chapter" data-level="4.3" data-path="working-with-large-data.html"><a href="working-with-large-data.html#data-download-and-import"><i class="fa fa-check"></i><b>4.3</b> Data download and import</a></li>
<li class="chapter" data-level="4.4" data-path="working-with-large-data.html"><a href="working-with-large-data.html#delayedarray-representation-of-large-on-disk-data"><i class="fa fa-check"></i><b>4.4</b> ‘DelayedArray’ representation of large on-disk data</a></li>
<li class="chapter" data-level="4.5" data-path="working-with-large-data.html"><a href="working-with-large-data.html#subseting-and-in-memory-location-for-data-exploration"><i class="fa fa-check"></i><b>4.5</b> Subseting and in-memory location for data exploration</a></li>
<li class="chapter" data-level="4.6" data-path="working-with-large-data.html"><a href="working-with-large-data.html#exploration-of-column-cell-annotations"><i class="fa fa-check"></i><b>4.6</b> Exploration of column (cell) annotations</a></li>
<li class="chapter" data-level="4.7" data-path="working-with-large-data.html"><a href="working-with-large-data.html#invariant-cell-annotations"><i class="fa fa-check"></i><b>4.7</b> Invariant cell annotations</a></li>
<li class="chapter" data-level="4.8" data-path="working-with-large-data.html"><a href="working-with-large-data.html#donor-provenance-and-organ-labels"><i class="fa fa-check"></i><b>4.8</b> Donor provenance and organ labels</a></li>
<li class="chapter" data-level="4.9" data-path="working-with-large-data.html"><a href="working-with-large-data.html#sub-setting-data-based-on-cell-attributes"><i class="fa fa-check"></i><b>4.9</b> Sub-setting data based on cell attributes</a></li>
<li class="chapter" data-level="4.10" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#information-about-the-packages-used-in-this-session"><i class="fa fa-check"></i><b>4.10</b> Information about the packages used in this session</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html"><i class="fa fa-check"></i><b>5</b> The AnVIL R / Bioconductor package</a><ul>
<li class="chapter" data-level="5.1" data-path="working-with-large-data.html"><a href="working-with-large-data.html#r-bioconductor-libraries"><i class="fa fa-check"></i><b>5.1</b> <em>R / Bioconductor</em> libraries</a></li>
<li class="chapter" data-level="5.2" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#the-anvil-workspace"><i class="fa fa-check"></i><b>5.2</b> The <em>AnVIL</em> workspace</a></li>
<li class="chapter" data-level="5.3" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#saving-analysis-products"><i class="fa fa-check"></i><b>5.3</b> Saving analysis products</a><ul>
<li class="chapter" data-level="5.3.1" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#an-abbreviated-work-flow"><i class="fa fa-check"></i><b>5.3.1</b> An abbreviated work flow</a></li>
<li class="chapter" data-level="5.3.2" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#saving-results-on-the-runtime-instance"><i class="fa fa-check"></i><b>5.3.2</b> Saving results on the runtime instance</a></li>
<li class="chapter" data-level="5.3.3" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#backing-up-results-to-the-workspace-bucket"><i class="fa fa-check"></i><b>5.3.3</b> Backing up results to the workspace bucket</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#workspace-data-tables"><i class="fa fa-check"></i><b>5.4</b> Workspace DATA ‘TABLES’</a><ul>
<li class="chapter" data-level="5.4.1" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#metadata-discovery-and-export-from-the-hca"><i class="fa fa-check"></i><b>5.4.1</b> Metadata discovery and export from the HCA</a></li>
<li class="chapter" data-level="5.4.2" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#accessing-tables"><i class="fa fa-check"></i><b>5.4.2</b> Accessing TABLES</a></li>
<li class="chapter" data-level="5.4.3" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#what-do-we-learn-about-the-samples-in-this-experiment"><i class="fa fa-check"></i><b>5.4.3</b> What do we learn about the samples in this experiment?</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#information-about-the-packages-used-in-this-session"><i class="fa fa-check"></i><b>5.5</b> Information about the packages used in this session</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="formal-work-flows-human-cell-atlas-optimus-pipeline.html"><a href="formal-work-flows-human-cell-atlas-optimus-pipeline.html"><i class="fa fa-check"></i><b>6</b> Formal work flows: Human Cell Atlas ‘Optimus’ pipeline</a><ul>
<li class="chapter" data-level="6.1" data-path="formal-work-flows-human-cell-atlas-optimus-pipeline.html"><a href="formal-work-flows-human-cell-atlas-optimus-pipeline.html#the-hca-optimus-single-cell-analysis-pipeline"><i class="fa fa-check"></i><b>6.1</b> The HCA Optimus single cell analysis pipeline</a></li>
<li class="chapter" data-level="6.2" data-path="formal-work-flows-human-cell-atlas-optimus-pipeline.html"><a href="formal-work-flows-human-cell-atlas-optimus-pipeline.html#running-a-work-flow"><i class="fa fa-check"></i><b>6.2</b> Running a work flow</a></li>
<li class="chapter" data-level="6.3" data-path="formal-work-flows-human-cell-atlas-optimus-pipeline.html"><a href="formal-work-flows-human-cell-atlas-optimus-pipeline.html#using-data-in-r"><i class="fa fa-check"></i><b>6.3</b> Using data in <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="next-steps.html"><a href="next-steps.html"><i class="fa fa-check"></i><b>7</b> Next steps</a><ul>
<li class="chapter" data-level="7.1" data-path="next-steps.html"><a href="next-steps.html#shiny-apps"><i class="fa fa-check"></i><b>7.1</b> Shiny apps</a></li>
<li class="chapter" data-level="7.2" data-path="next-steps.html"><a href="next-steps.html#protected-data-access"><i class="fa fa-check"></i><b>7.2</b> Protected data access</a></li>
<li class="chapter" data-level="7.3" data-path="next-steps.html"><a href="next-steps.html#other-clouds"><i class="fa fa-check"></i><b>7.3</b> Other clouds</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R / Bioconductor in the AnVIL Cloud</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="the-anvil-r-bioconductor-package" class="section level1">
<h1><span class="header-section-number">Chapter 5</span> The AnVIL R / Bioconductor package</h1>
<p>In this section we learn about</p>
<ul>
<li>Saving data to the current runtime</li>
<li>Saving or retrieving data from the workspace bucket</li>
<li>Manipulating the ‘TABLES’ metadata</li>
<li>‘Power-user’ access to the underlying AnVIL software components</li>
</ul>
<p>Packages used include</p>
<ul>
<li><a href="https://data.humancellatlas.org/explore/projects">AnVIL</a> – access AnVIL and Google cloud resources from within
R. The <code>AnVIL</code> package is under active development, and we use the
most recent version installed from Github.</li>
</ul>
<div id="r-bioconductor-libraries" class="section level2">
<h2><span class="header-section-number">5.1</span> <em>R / Bioconductor</em> libraries</h2>
<p>We’ll primarily use the <code>AnVIL</code> package, installed from github.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">github_packages &lt;-<span class="st"> &quot;Bioconductor/AnVIL&quot;</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">idx &lt;-<span class="st"> </span><span class="op">!</span><span class="st"> </span><span class="kw">basename</span>(github_packages) <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>())</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">need &lt;-<span class="st"> </span>github_packages[ idx ]</a>
<a class="sourceLine" id="cb1-4" data-line-number="4">BiocManager<span class="op">::</span><span class="kw">install</span>(need, <span class="dt">update =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>## Bioconductor version 3.11 (BiocManager 1.30.10), R 4.0.2 Patched (2020-06-24
##   r78747)</code></pre>
<p>We’ll motivate the use of <code>AnVIL</code> using a previous work flow, and use
of the <code>dplyr</code> package. The required packages can be installed with</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">packages &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;scRNAseq&quot;</span>, <span class="st">&quot;scater&quot;</span>, <span class="st">&quot;scran&quot;</span>, <span class="st">&quot;uwot&quot;</span>, <span class="st">&quot;dplyr&quot;</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">need &lt;-<span class="st"> </span>packages[<span class="op">!</span>packages <span class="op">%in%</span><span class="st"> </span><span class="kw">rownames</span>(<span class="kw">installed.packages</span>())]</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">BiocManager<span class="op">::</span><span class="kw">install</span>(need, <span class="dt">update =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>## Bioconductor version 3.11 (BiocManager 1.30.10), R 4.0.2 Patched (2020-06-24
##   r78747)</code></pre>
<p>Start by loading the <code>AnVIL</code> package.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">suppressPackageStartupMessages</span>({</a>
<a class="sourceLine" id="cb5-2" data-line-number="2">    <span class="kw">library</span>(AnVIL)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">})</a></code></pre></div>
</div>
<div id="the-anvil-workspace" class="section level2">
<h2><span class="header-section-number">5.2</span> The <em>AnVIL</em> workspace</h2>
<p>The central components of the AnVIL workspace are available. Each
workspace has a ‘namespace’ (billing account) and ‘name’ (workspace
name).</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">avworkspace_namespace</span>() <span class="co"># billing account</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw">avworkspace_name</span>()      <span class="co"># workspace name</span></a></code></pre></div>
<p>Each workspace has a google ‘bucket’ associated with it. The bucket is
in existence for as long as the workspace, providing an area for
persistent data storage across different kernels.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw">avbucket</span>()</a></code></pre></div>
<pre><code>## [1] &quot;gs://fc-66b30014-b9a7-4374-a24a-e9eba924b67d&quot;</code></pre>
</div>
<div id="saving-analysis-products" class="section level2">
<h2><span class="header-section-number">5.3</span> Saving analysis products</h2>
<div id="an-abbreviated-work-flow" class="section level3">
<h3><span class="header-section-number">5.3.1</span> An abbreviated work flow</h3>
<p>Here’s an abbreviated version of an earlier work flow</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co">## Necessary packages</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="kw">suppressPackageStartupMessages</span>({</a>
<a class="sourceLine" id="cb9-3" data-line-number="3">    <span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb9-4" data-line-number="4">    <span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb9-5" data-line-number="5">    <span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb9-6" data-line-number="6">})</a>
<a class="sourceLine" id="cb9-7" data-line-number="7"></a>
<a class="sourceLine" id="cb9-8" data-line-number="8"><span class="co">## Pre-configured data</span></a>
<a class="sourceLine" id="cb9-9" data-line-number="9"><span class="kw">suppressMessages</span>({</a>
<a class="sourceLine" id="cb9-10" data-line-number="10">    sce &lt;-<span class="st"> </span><span class="kw">MacoskoRetinaData</span>()</a>
<a class="sourceLine" id="cb9-11" data-line-number="11">})</a>
<a class="sourceLine" id="cb9-12" data-line-number="12"></a>
<a class="sourceLine" id="cb9-13" data-line-number="13"><span class="co">## QC</span></a>
<a class="sourceLine" id="cb9-14" data-line-number="14">is.mito &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;^MT-&quot;</span>, <span class="kw">rownames</span>(sce))</a>
<a class="sourceLine" id="cb9-15" data-line-number="15">qcstats &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(sce, <span class="dt">subsets=</span><span class="kw">list</span>(<span class="dt">Mito=</span>is.mito))</a>
<a class="sourceLine" id="cb9-16" data-line-number="16">filtered &lt;-<span class="st"> </span><span class="kw">quickPerCellQC</span>(qcstats, <span class="dt">percent_subsets=</span><span class="st">&quot;subsets_Mito_percent&quot;</span>)</a>
<a class="sourceLine" id="cb9-17" data-line-number="17">sce &lt;-<span class="st"> </span>sce[, <span class="op">!</span>filtered<span class="op">$</span>discard]</a>
<a class="sourceLine" id="cb9-18" data-line-number="18"></a>
<a class="sourceLine" id="cb9-19" data-line-number="19"><span class="co">## Normalization</span></a>
<a class="sourceLine" id="cb9-20" data-line-number="20">sce &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sce)</a>
<a class="sourceLine" id="cb9-21" data-line-number="21"></a>
<a class="sourceLine" id="cb9-22" data-line-number="22"><span class="co">## Filtering</span></a>
<a class="sourceLine" id="cb9-23" data-line-number="23">dec &lt;-<span class="st"> </span><span class="kw">modelGeneVar</span>(sce)</a>
<a class="sourceLine" id="cb9-24" data-line-number="24">hvg &lt;-<span class="st"> </span><span class="kw">getTopHVGs</span>(dec, <span class="dt">prop=</span><span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb9-25" data-line-number="25"></a>
<a class="sourceLine" id="cb9-26" data-line-number="26"><span class="co">## Dimensionality reduction</span></a>
<a class="sourceLine" id="cb9-27" data-line-number="27"><span class="kw">set.seed</span>(<span class="dv">1234</span>)</a>
<a class="sourceLine" id="cb9-28" data-line-number="28">sce &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce, <span class="dt">ncomponents=</span><span class="dv">25</span>, <span class="dt">subset_row=</span>hvg)</a>
<a class="sourceLine" id="cb9-29" data-line-number="29">sce &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce, <span class="dt">dimred =</span> <span class="st">&#39;PCA&#39;</span>, <span class="dt">external_neighbors=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb9-30" data-line-number="30"></a>
<a class="sourceLine" id="cb9-31" data-line-number="31"><span class="co">## Clustering</span></a>
<a class="sourceLine" id="cb9-32" data-line-number="32">g &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(sce, <span class="dt">use.dimred =</span> <span class="st">&#39;PCA&#39;</span>)</a>
<a class="sourceLine" id="cb9-33" data-line-number="33"><span class="kw">colLabels</span>(sce) &lt;-<span class="st"> </span><span class="kw">factor</span>(igraph<span class="op">::</span><span class="kw">cluster_louvain</span>(g)<span class="op">$</span>membership)</a></code></pre></div>
</div>
<div id="saving-results-on-the-runtime-instance" class="section level3">
<h3><span class="header-section-number">5.3.2</span> Saving results on the runtime instance</h3>
<p>At this point, we’ve invested a certain amount of effort, mental and
computational, into producing the updated <code>sce</code>. It might be valuable
to save this object so that we can quickly read it in during our next
session.</p>
<p>The standard mechanism for saving an <em>R</em> object might use <code>saveRDS()</code>
to write the image to disk, and <code>readRDS()</code> to load it back in:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co">## save the file in a temporary directory, for illustration purposes</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2">tmp_dir &lt;-<span class="st"> </span><span class="kw">tempfile</span>(); <span class="kw">dir.create</span>(tmp_dir)</a>
<a class="sourceLine" id="cb10-3" data-line-number="3">file_name &lt;-<span class="st"> </span><span class="kw">file.path</span>(tmp_dir, <span class="kw">format</span>(<span class="kw">Sys.Date</span>(), <span class="st">&quot;sce_%m-%d-%y.RDS&quot;</span>))</a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="kw">saveRDS</span>(sce, file_name)</a>
<a class="sourceLine" id="cb10-5" data-line-number="5"></a>
<a class="sourceLine" id="cb10-6" data-line-number="6">sce1 &lt;-<span class="st"> </span><span class="kw">readRDS</span>(file_name)</a>
<a class="sourceLine" id="cb10-7" data-line-number="7">sce1</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 24658 45877 
## metadata(0):
## assays(2): counts logcounts
## rownames(24658): KITL TMTC3 ... 1110059M19RIK GM20861
## rowData names(0):
## colnames(45877): r1_GGCCGCAGTCCG r1_CTTGTGCGGGAA ... p1_TAACGCGCTCCT
##   p1_ATTCTTGTTCTT
## colData names(4): cell.id cluster sizeFactor label
## reducedDimNames(2): PCA UMAP
## altExpNames(0):</code></pre>
</div>
<div id="backing-up-results-to-the-workspace-bucket" class="section level3">
<h3><span class="header-section-number">5.3.3</span> Backing up results to the workspace bucket</h3>
<p>A challenge with this approach is that the file system is associated
with the runtime. When the runtime ends, intentionally (e.g., to
switch to a different runtime) or accidentally (e.g., because of some
technical problem in the AnVIL), the file is no longer available.</p>
<p>To make ‘persist’ the file so that its lifespan extends to the life of
the <em>workspace</em>, copy the file from it’s current location to the
google bucket associated with the workspace. Do this using
<code>avfiles_backup()</code> (<code>res</code> contains the output of the cloud copy
command, and would be useful if debugging a surprising outcome).</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">res &lt;-<span class="st"> </span><span class="kw">avfiles_backup</span>(file_name)</a></code></pre></div>
<p>The content of the google bucket can be viewed with</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="kw">print</span>( <span class="kw">avfiles_ls</span>(<span class="dt">recursive =</span> <span class="ot">TRUE</span>) )</a></code></pre></div>
<pre><code>## [1] &quot;sce_07-16-20.RDS&quot;                     
## [2] &quot;notebooks/03-existing-workspace.ipynb&quot;
## [3] &quot;notebooks/04-hca-cloud-R.ipynb&quot;       
## [4] &quot;notebooks/05-AnVIL-package.ipynb&quot;</code></pre>
<p>We see that the data set, and the notebooks, are stored in the
bucket. The file can be restored to the runtime, perhaps in a
different location, with</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">tmp &lt;-<span class="st"> </span><span class="kw">tempfile</span>(); <span class="kw">dir.create</span>(tmp)</a>
<a class="sourceLine" id="cb15-2" data-line-number="2"><span class="kw">print</span>( <span class="kw">dir</span>(tmp) ) <span class="co"># nothing there...</span></a></code></pre></div>
<pre><code>## character(0)</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">res &lt;-<span class="kw">avfiles_restore</span>(<span class="kw">basename</span>(file_name), tmp)</a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="kw">print</span>( <span class="kw">dir</span>(tmp) ) <span class="co"># ...now there is!</span></a></code></pre></div>
<pre><code>## [1] &quot;sce_07-16-20.RDS&quot;</code></pre>
<p>This restoration works even if the runtime has been changed, e.g., to
use RStudio instead of Jupyter notebooks.</p>
</div>
</div>
<div id="workspace-data-tables" class="section level2">
<h2><span class="header-section-number">5.4</span> Workspace DATA ‘TABLES’</h2>
<p>Check out the DATA tab on the workspace home page. Note that there is
a TABLES entry. A Workspace data TABLE provides metadata about
genome-scale resources available to the workspace. Often, the data are
subject to access restrictions, and access to the workspace implies
access to the restricted data.</p>
<p>We’ll create references to public-access data available from the Human
Cell Atlas.</p>
<div id="metadata-discovery-and-export-from-the-hca" class="section level3">
<h3><span class="header-section-number">5.4.1</span> Metadata discovery and export from the HCA</h3>
<p>Visit the <a href="https://data.humancellatlas.org/explore/projects">Human Cell Atlas Data Portal</a>.</p>
<p>Select on the project “<a href="https://data.humancellatlas.org/explore/projects/4a95101c-9ffc-4f30-a809-f04518a23803">A single-cell reference map of transcriptional
states for human blood and tissue T cell activation</a>” (second item
on the default screen, as of this writing). Note that the data set is
very well documented. We worked with the ‘Expression Matrices’ (left
panel) in a previous section of the workshop, and will work with
‘Analysis Protocol (optimus_v1.3.5)’ (right sidebar) in a subsequent
section.</p>
<p>Return to the <a href="https://data.humancellatlas.org/explore/projects">data portal</a>. Use the check box to select the
project, and click ‘Export Selected Data’ toward the top right.</p>
<p>On the next screen, choose ‘Export to Terra’, select ‘BAM’ files, and
‘Request Export’</p>
<p>I was transferred to Terra, asked to log in, then to create or use a
workspace. I chose an existing workspace (this one!). The HCA data
appeared under the DATA tab as a TABLE labeled “participants”.</p>
<p>The AnVIL package allow us to query the workspace data tables.</p>
</div>
<div id="accessing-tables" class="section level3">
<h3><span class="header-section-number">5.4.2</span> Accessing TABLES</h3>
<p>Use <code>avtables()</code> to discover the tables available.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw">avtables</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="st">    </span><span class="kw">print</span>()</a></code></pre></div>
<pre><code>## # A tibble: 1 x 3
##   table      count colnames                                                     
##   &lt;chr&gt;      &lt;int&gt; &lt;chr&gt;                                                        
## 1 participa…    16 participant_id, bundle_uuid, bundle_version, cell_suspension…</code></pre>
<p>Use <code>avtable()</code> to retrieve a particular table.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">participant &lt;-<span class="st"> </span><span class="kw">avtable</span>(<span class="st">&quot;participant&quot;</span>)</a>
<a class="sourceLine" id="cb21-2" data-line-number="2"><span class="kw">print</span>(participant)</a></code></pre></div>
<pre><code>## # A tibble: 16 x 121
##    name  `__unknown_3__f… `__csv__file_na… `__unknown_4__f… donor_organism_…
##    &lt;chr&gt; &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;           
##  1 09ce… unknown          empty_drops_res… 60921956ce081e3… 95822e01-44a7-4…
##  2 0cba… unknown          empty_drops_res… bf8fbfd77dc4dfe… b09a1e10-3c38-4…
##  3 24cf… unknown          empty_drops_res… 00ccbe66273c95e… 95822e01-44a7-4…
##  4 2677… unknown          empty_drops_res… 4cc66da5a4279cd… 8b5774b2-2aba-4…
##  5 269c… unknown          empty_drops_res… 823bb11d2b3d5f8… 8b5774b2-2aba-4…
##  6 6aeb… unknown          empty_drops_res… 004b9ce987100ed… 8b5774b2-2aba-4…
##  7 88f0… unknown          empty_drops_res… 3fb281d48b034bc… 8b5774b2-2aba-4…
##  8 8b9e… unknown          empty_drops_res… 048b463622830aa… 8b5774b2-2aba-4…
##  9 8c6e… unknown          empty_drops_res… b55f255b6892ab4… 95822e01-44a7-4…
## 10 9137… unknown          empty_drops_res… bdcac1fa24a6af8… 95822e01-44a7-4…
## 11 a460… unknown          empty_drops_res… aed8ac6e3611608… 95822e01-44a7-4…
## 12 c4dc… unknown          empty_drops_res… 91248e040be8119… b09a1e10-3c38-4…
## 13 c5df… unknown          empty_drops_res… 872c8366048261e… 8b5774b2-2aba-4…
## 14 cfa6… unknown          empty_drops_res… 351bf8067ef3b42… 4b7a3de7-94db-4…
## 15 dde6… unknown          empty_drops_res… bd41fb241c3b618… 95822e01-44a7-4…
## 16 ecbc… unknown          empty_drops_res… a0598691e0a040f… 4b7a3de7-94db-4…
## # … with 116 more variables: `__unknown_3__file_sha256` &lt;chr&gt;,
## #   `__fastq_read2__file_sha256` &lt;chr&gt;, `__unknown_0__file_size` &lt;chr&gt;,
## #   `__fastq_read2__drs_url` &lt;chr&gt;, `__unknown_0__file_sha256` &lt;chr&gt;,
## #   `__csv__drs_url` &lt;chr&gt;, `__fastq_read2__file_format` &lt;chr&gt;,
## #   `__unknown_1__file_version` &lt;chr&gt;, `__fastq_read2__file_name` &lt;chr&gt;,
## #   cell_suspension__estimated_cell_count &lt;chr&gt;, `__unknown_4__drs_url` &lt;chr&gt;,
## #   `__bam__file_url` &lt;chr&gt;, `__unknown_2__file_uuid` &lt;chr&gt;,
## #   `__unknown_3__file_name` &lt;chr&gt;, `__unknown_1__file_content_type` &lt;chr&gt;,
## #   `__unknown_0__file_version` &lt;chr&gt;, `__bam__file_sha256` &lt;chr&gt;,
## #   `__csv__file_sha256` &lt;chr&gt;, `__bam__file_version` &lt;chr&gt;,
## #   cell_suspension__selected_cell_type &lt;chr&gt;,
## #   `__fastq_read1__file_name` &lt;chr&gt;,
## #   sequencing_process__provenance__document_id &lt;chr&gt;,
## #   `__matrix__file_content_type` &lt;chr&gt;,
## #   `__unknown_3__file_content_type` &lt;chr&gt;,
## #   specimen_from_organism__provenance__document_id &lt;chr&gt;,
## #   project__contributors__laboratory &lt;chr&gt;, `__matrix__file_name` &lt;chr&gt;,
## #   sequencing_protocol__paired_end &lt;chr&gt;, `__unknown_4__file_format` &lt;chr&gt;,
## #   `__unknown_2__file_format` &lt;chr&gt;, `__unknown_4__file_uuid` &lt;chr&gt;,
## #   `__unknown_1__file_name` &lt;chr&gt;, sample__provenance__document_id &lt;chr&gt;,
## #   `_entity_type` &lt;chr&gt;, `__unknown_0__drs_url` &lt;chr&gt;,
## #   `__fastq_read2__file_version` &lt;chr&gt;, `__bam__file_content_type` &lt;chr&gt;,
## #   `__unknown_0__file_format` &lt;chr&gt;, `__fastq_read1__file_size` &lt;chr&gt;,
## #   donor_organism__organism_age &lt;chr&gt;, `__unknown_1__file_uuid` &lt;chr&gt;,
## #   project__project_core__project_short_name &lt;chr&gt;,
## #   `__unknown_2__file_url` &lt;chr&gt;, donor_organism__diseases &lt;chr&gt;,
## #   bundle_version &lt;chr&gt;, `__csv__file_uuid` &lt;chr&gt;,
## #   `__unknown_4__file_content_type` &lt;chr&gt;,
## #   library_preparation_protocol__library_construction_approach &lt;chr&gt;,
## #   `__matrix__drs_url` &lt;chr&gt;, `__bam__file_uuid` &lt;chr&gt;,
## #   `__unknown_4__file_version` &lt;chr&gt;, `__unknown_0__file_url` &lt;chr&gt;,
## #   `__fastq_read2__file_content_type` &lt;chr&gt;, `__bam__file_size` &lt;chr&gt;,
## #   `__matrix__file_url` &lt;chr&gt;, `__unknown_2__file_sha256` &lt;chr&gt;,
## #   `__fastq_read2__read_index` &lt;chr&gt;, `__csv__file_version` &lt;chr&gt;,
## #   `__matrix__file_version` &lt;chr&gt;, `__bam__drs_url` &lt;chr&gt;,
## #   `__csv__file_format` &lt;chr&gt;,
## #   sequencing_protocol__instrument_manufacturer_model &lt;chr&gt;,
## #   `__fastq_read1__file_content_type` &lt;chr&gt;, `__unknown_3__file_uuid` &lt;chr&gt;,
## #   bundle_uuid &lt;chr&gt;, project__provenance__document_id &lt;chr&gt;,
## #   cell_suspension__provenance__document_id &lt;chr&gt;,
## #   `__unknown_3__file_size` &lt;chr&gt;, `__unknown_3__file_version` &lt;chr&gt;,
## #   `__unknown_4__file_size` &lt;chr&gt;, `__unknown_2__file_name` &lt;chr&gt;,
## #   sample__biomaterial_core__biomaterial_id &lt;chr&gt;,
## #   `__fastq_read1__file_format` &lt;chr&gt;, donor_organism__sex &lt;chr&gt;,
## #   `__unknown_4__file_name` &lt;chr&gt;, `__unknown_0__file_name` &lt;chr&gt;,
## #   `__fastq_read1__file_sha256` &lt;chr&gt;, `__matrix__file_sha256` &lt;chr&gt;,
## #   `__matrix__file_format` &lt;chr&gt;, donor_organism__genus_species &lt;chr&gt;,
## #   `__unknown_2__drs_url` &lt;chr&gt;, `__unknown_1__file_size` &lt;chr&gt;,
## #   `__fastq_read1__file_version` &lt;chr&gt;, `__unknown_3__drs_url` &lt;chr&gt;,
## #   `__unknown_3__file_url` &lt;chr&gt;, project__project_core__project_title &lt;chr&gt;,
## #   `__fastq_read1__read_index` &lt;chr&gt;, `__fastq_read2__file_size` &lt;chr&gt;,
## #   `__matrix__file_size` &lt;chr&gt;, `__fastq_read1__file_uuid` &lt;chr&gt;,
## #   `__unknown_1__file_format` &lt;chr&gt;, `__unknown_1__file_url` &lt;chr&gt;,
## #   `__csv__file_size` &lt;chr&gt;,
## #   specimen_from_organism__preservation_storage__preservation_method &lt;chr&gt;,
## #   `__unknown_2__file_content_type` &lt;chr&gt;, `__bam__file_format` &lt;chr&gt;,
## #   donor_organism__biomaterial_core__biomaterial_id &lt;chr&gt;,
## #   `__bam__file_name` &lt;chr&gt;, `__unknown_1__file_sha256` &lt;chr&gt;,
## #   `__unknown_1__drs_url` &lt;chr&gt;, …</code></pre>
</div>
<div id="what-do-we-learn-about-the-samples-in-this-experiment" class="section level3">
<h3><span class="header-section-number">5.4.3</span> What do we learn about the samples in this experiment?</h3>
<p>Metadata about the participants in the study can be found by selecting
the columns that do not start with an underscore</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">participant <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb23-2" data-line-number="2"><span class="st">    </span><span class="kw">select</span>( <span class="op">!</span><span class="kw">starts_with</span>(<span class="st">&quot;_&quot;</span>) ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb23-3" data-line-number="3"><span class="st">    </span><span class="kw">print</span>()</a></code></pre></div>
<pre><code>## # A tibble: 16 x 28
##    name  donor_organism_… cell_suspension… cell_suspension… sequencing_proc…
##    &lt;chr&gt; &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;           
##  1 09ce… 95822e01-44a7-4… 0                T cell           294fe5d9-c1e8-4…
##  2 0cba… b09a1e10-3c38-4… 0                T cell           bfbf2ca6-13e5-4…
##  3 24cf… 95822e01-44a7-4… 0                T cell           a040dae6-e0b1-4…
##  4 2677… 8b5774b2-2aba-4… 0                T cell           d6536459-ab4e-4…
##  5 269c… 8b5774b2-2aba-4… 0                T cell           58cf25fd-440a-4…
##  6 6aeb… 8b5774b2-2aba-4… 0                T cell           098cc66a-d806-4…
##  7 88f0… 8b5774b2-2aba-4… 0                T cell           c76d90b8-c190-4…
##  8 8b9e… 8b5774b2-2aba-4… 0                T cell           c763f679-e13d-4…
##  9 8c6e… 95822e01-44a7-4… 0                T cell           58a18a4c-5423-4…
## 10 9137… 95822e01-44a7-4… 0                T cell           36ca61c4-9752-4…
## 11 a460… 95822e01-44a7-4… 0                T cell           6fcd2c86-76dc-4…
## 12 c4dc… b09a1e10-3c38-4… 0                T cell           219e1b92-9749-4…
## 13 c5df… 8b5774b2-2aba-4… 0                T cell           fcbaa3c6-b2bf-4…
## 14 cfa6… 4b7a3de7-94db-4… 0                T cell           24ae6c0b-d147-4…
## 15 dde6… 95822e01-44a7-4… 0                T cell           fb72f4c2-7f35-4…
## 16 ecbc… 4b7a3de7-94db-4… 0                T cell           3ddf143f-36bd-4…
## # … with 23 more variables:
## #   specimen_from_organism__provenance__document_id &lt;chr&gt;,
## #   project__contributors__laboratory &lt;chr&gt;,
## #   sequencing_protocol__paired_end &lt;chr&gt;,
## #   sample__provenance__document_id &lt;chr&gt;, donor_organism__organism_age &lt;chr&gt;,
## #   project__project_core__project_short_name &lt;chr&gt;,
## #   donor_organism__diseases &lt;chr&gt;, bundle_version &lt;chr&gt;,
## #   library_preparation_protocol__library_construction_approach &lt;chr&gt;,
## #   sequencing_protocol__instrument_manufacturer_model &lt;chr&gt;,
## #   bundle_uuid &lt;chr&gt;, project__provenance__document_id &lt;chr&gt;,
## #   cell_suspension__provenance__document_id &lt;chr&gt;,
## #   sample__biomaterial_core__biomaterial_id &lt;chr&gt;, donor_organism__sex &lt;chr&gt;,
## #   donor_organism__genus_species &lt;chr&gt;,
## #   project__project_core__project_title &lt;chr&gt;,
## #   specimen_from_organism__preservation_storage__preservation_method &lt;chr&gt;,
## #   donor_organism__biomaterial_core__biomaterial_id &lt;chr&gt;,
## #   specimen_from_organism__organ_part &lt;chr&gt;,
## #   donor_organism__organism_age_unit &lt;chr&gt;,
## #   project__contributors__institution &lt;chr&gt;,
## #   specimen_from_organism__organ &lt;chr&gt;</code></pre>
<p>As in a previous exercise, we can select the invariant columns to learn
about experiment-wide participant descriptions…</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1">participant <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb25-2" data-line-number="2"><span class="st">    </span><span class="kw">select</span>( <span class="op">!</span><span class="kw">starts_with</span>(<span class="st">&quot;_&quot;</span>) ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb25-3" data-line-number="3"><span class="st">    </span><span class="kw">select_if</span>( <span class="op">~</span><span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(.)) <span class="op">==</span><span class="st"> </span>1L) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb25-4" data-line-number="4"><span class="st">    </span><span class="kw">distinct</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb25-5" data-line-number="5"><span class="st">    </span>tidyr<span class="op">::</span><span class="kw">pivot_longer</span>(<span class="kw">everything</span>()) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb25-6" data-line-number="6"><span class="st">    </span><span class="kw">print</span>()</a></code></pre></div>
<pre><code>## # A tibble: 14 x 2
##    name                                value                                    
##    &lt;chr&gt;                               &lt;chr&gt;                                    
##  1 cell_suspension__estimated_cell_co… 0                                        
##  2 cell_suspension__selected_cell_type T cell                                   
##  3 project__contributors__laboratory   Farber Lab; Columbia Center for Translat…
##  4 sequencing_protocol__paired_end     False                                    
##  5 project__project_core__project_sho… HumanTissueTcellActivation               
##  6 library_preparation_protocol__libr… 10X v2 sequencing                        
##  7 sequencing_protocol__instrument_ma… Illumina HiSeq 4000                      
##  8 project__provenance__document_id    4a95101c-9ffc-4f30-a809-f04518a23803     
##  9 donor_organism__sex                 male                                     
## 10 donor_organism__genus_species       Homo sapiens                             
## 11 project__project_core__project_tit… A single-cell reference map of transcrip…
## 12 specimen_from_organism__preservati… fresh                                    
## 13 donor_organism__organism_age_unit   year                                     
## 14 project__contributors__institution  Columbia University Irving Medical Cente…</code></pre>
<p>The varying columns…</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1">participant <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-2" data-line-number="2"><span class="st">    </span><span class="kw">select</span>( <span class="op">!</span><span class="kw">starts_with</span>(<span class="st">&quot;_&quot;</span>) ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-3" data-line-number="3"><span class="st">    </span><span class="kw">select_if</span>( <span class="op">~</span><span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(.)) <span class="op">!=</span><span class="st"> </span>1L) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb27-4" data-line-number="4"><span class="st">    </span><span class="kw">print</span>()</a></code></pre></div>
<pre><code>## # A tibble: 16 x 14
##    name  donor_organism_… sequencing_proc… specimen_from_o… sample__provena…
##    &lt;chr&gt; &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;            &lt;chr&gt;           
##  1 09ce… 95822e01-44a7-4… 294fe5d9-c1e8-4… c9abc146-67f0-4… c9abc146-67f0-4…
##  2 0cba… b09a1e10-3c38-4… bfbf2ca6-13e5-4… 0ec78ea7-1fc1-4… 0ec78ea7-1fc1-4…
##  3 24cf… 95822e01-44a7-4… a040dae6-e0b1-4… ed854de5-c872-4… ed854de5-c872-4…
##  4 2677… 8b5774b2-2aba-4… d6536459-ab4e-4… dfe889d8-86c0-4… dfe889d8-86c0-4…
##  5 269c… 8b5774b2-2aba-4… 58cf25fd-440a-4… c2fed28d-cdd4-4… c2fed28d-cdd4-4…
##  6 6aeb… 8b5774b2-2aba-4… 098cc66a-d806-4… a3c3f746-aa00-4… a3c3f746-aa00-4…
##  7 88f0… 8b5774b2-2aba-4… c76d90b8-c190-4… 427e0abb-3d3e-4… 427e0abb-3d3e-4…
##  8 8b9e… 8b5774b2-2aba-4… c763f679-e13d-4… 55c59b93-6ce2-4… 55c59b93-6ce2-4…
##  9 8c6e… 95822e01-44a7-4… 58a18a4c-5423-4… 9a79497c-a825-4… 9a79497c-a825-4…
## 10 9137… 95822e01-44a7-4… 36ca61c4-9752-4… 85a0036b-fb11-4… 85a0036b-fb11-4…
## 11 a460… 95822e01-44a7-4… 6fcd2c86-76dc-4… 59a13e80-0c0b-4… 59a13e80-0c0b-4…
## 12 c4dc… b09a1e10-3c38-4… 219e1b92-9749-4… 46cbd6a3-1ba4-4… 46cbd6a3-1ba4-4…
## 13 c5df… 8b5774b2-2aba-4… fcbaa3c6-b2bf-4… d02c3fdd-5c85-4… d02c3fdd-5c85-4…
## 14 cfa6… 4b7a3de7-94db-4… 24ae6c0b-d147-4… edd28ba8-0bb3-4… edd28ba8-0bb3-4…
## 15 dde6… 95822e01-44a7-4… fb72f4c2-7f35-4… ef3a770b-6bd5-4… ef3a770b-6bd5-4…
## 16 ecbc… 4b7a3de7-94db-4… 3ddf143f-36bd-4… 03a73511-bdeb-4… 03a73511-bdeb-4…
## # … with 9 more variables: donor_organism__organism_age &lt;chr&gt;,
## #   donor_organism__diseases &lt;chr&gt;, bundle_version &lt;chr&gt;, bundle_uuid &lt;chr&gt;,
## #   cell_suspension__provenance__document_id &lt;chr&gt;,
## #   sample__biomaterial_core__biomaterial_id &lt;chr&gt;,
## #   donor_organism__biomaterial_core__biomaterial_id &lt;chr&gt;,
## #   specimen_from_organism__organ_part &lt;chr&gt;,
## #   specimen_from_organism__organ &lt;chr&gt;</code></pre>
<p>Details of, e.g., donor attributes are easily discovered</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1">donor &lt;-</a>
<a class="sourceLine" id="cb29-2" data-line-number="2"><span class="st">    </span>participant <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb29-3" data-line-number="3"><span class="st">    </span><span class="kw">select</span>( <span class="kw">starts_with</span>(<span class="st">&quot;donor_organism__&quot;</span>) ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb29-4" data-line-number="4"><span class="st">    </span><span class="co">## for visual display, drop the &#39;donor_organism__&#39; prefix</span></a>
<a class="sourceLine" id="cb29-5" data-line-number="5"><span class="st">    </span><span class="kw">rename_all</span>( <span class="op">~</span><span class="st"> </span><span class="kw">sub</span>(<span class="st">&quot;donor_organism__&quot;</span>, <span class="st">&quot;&quot;</span>, .) )</a>
<a class="sourceLine" id="cb29-6" data-line-number="6"></a>
<a class="sourceLine" id="cb29-7" data-line-number="7">donor <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb29-8" data-line-number="8"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw">count</span>(</a>
<a class="sourceLine" id="cb29-9" data-line-number="9">               provenance__document_id,</a>
<a class="sourceLine" id="cb29-10" data-line-number="10">               sex,</a>
<a class="sourceLine" id="cb29-11" data-line-number="11">               organism_age</a>
<a class="sourceLine" id="cb29-12" data-line-number="12">           ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb29-13" data-line-number="13"><span class="st">    </span><span class="kw">print</span>()</a></code></pre></div>
<pre><code>## # A tibble: 4 x 4
##   provenance__document_id              sex   organism_age     n
##   &lt;chr&gt;                                &lt;chr&gt; &lt;chr&gt;        &lt;int&gt;
## 1 4b7a3de7-94db-4507-b1f9-c3931f9f3aa5 male  50-55            2
## 2 8b5774b2-2aba-4b90-8fa0-c7f69207802c male  65               6
## 3 95822e01-44a7-420b-92fc-b001460b1d13 male  52               6
## 4 b09a1e10-3c38-4bb3-8553-d762810a6fb7 male  50-55            2</code></pre>
<p>The ‘big data’ (e.g., BAM files) are co-located in the google cloud
(note the <code>replica=gcp</code>, i.e., Google cloud platform, at the end of
the URL), so accessible for computation and without download charge.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1">participant <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb31-2" data-line-number="2"><span class="st">    </span><span class="kw">select</span>( <span class="kw">contains</span>(<span class="st">&quot;bam__file_url&quot;</span>) )</a></code></pre></div>
<pre><code>## # A tibble: 16 x 1
##    `__bam__file_url`                                                            
##    &lt;chr&gt;                                                                        
##  1 https://dss.data.humancellatlas.org/v1/files/4ad1cc30-98e2-4db1-a303-20fb297…
##  2 https://dss.data.humancellatlas.org/v1/files/9af75eda-79b9-48b8-a29b-2d162cf…
##  3 https://dss.data.humancellatlas.org/v1/files/47d4aee5-06e9-4821-badf-08155c9…
##  4 https://dss.data.humancellatlas.org/v1/files/b8c092c8-4a89-45b6-bed3-4cc2bb5…
##  5 https://dss.data.humancellatlas.org/v1/files/cdb319d6-bd03-4984-a94d-c8bb3d6…
##  6 https://dss.data.humancellatlas.org/v1/files/048541ec-503a-48a0-85f8-58eba00…
##  7 https://dss.data.humancellatlas.org/v1/files/8424fdb4-c168-4cac-85d8-59694be…
##  8 https://dss.data.humancellatlas.org/v1/files/896605ca-71e8-449a-8cc4-1dd8bf0…
##  9 https://dss.data.humancellatlas.org/v1/files/b6b7f5a1-3f4b-4d88-8f1e-0f682b5…
## 10 https://dss.data.humancellatlas.org/v1/files/50897f4d-a010-4626-9547-92b8eaa…
## 11 https://dss.data.humancellatlas.org/v1/files/b9fa1e76-4999-4358-a18b-9b2cee2…
## 12 https://dss.data.humancellatlas.org/v1/files/ab102f17-5045-4b69-9cd3-2b22ca4…
## 13 https://dss.data.humancellatlas.org/v1/files/87a84844-349c-4463-8c0b-6aa22b1…
## 14 https://dss.data.humancellatlas.org/v1/files/f97266be-5f00-48de-b0cf-a3528fb…
## 15 https://dss.data.humancellatlas.org/v1/files/2802b4ef-88d0-4283-8d92-a2573eb…
## 16 https://dss.data.humancellatlas.org/v1/files/b04813fc-ac15-4fbb-8f14-9d8796a…</code></pre>
</div>
</div>
<div id="information-about-the-packages-used-in-this-session" class="section level2">
<h2><span class="header-section-number">5.5</span> Information about the packages used in this session</h2>
<p>The <em>R</em> command <code>sessionInfo()</code> captures information about the
versions of software used in the current session. This can be valuable
for performing reproducible analysis.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb33-1" data-line-number="1"><span class="kw">sessionInfo</span>()</a></code></pre></div>
<pre><code>## R version 4.0.2 Patched (2020-06-24 r78747)
## Platform: x86_64-apple-darwin17.7.0 (64-bit)
## Running under: macOS High Sierra 10.13.6
## 
## Matrix products: default
## BLAS:   /Users/ma38727/bin/R-4-0-branch/lib/libRblas.dylib
## LAPACK: /Users/ma38727/bin/R-4-0-branch/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] LoomExperiment_1.6.0        rtracklayer_1.48.0         
##  [3] rhdf5_2.32.2                BiocFileCache_1.12.0       
##  [5] dbplyr_1.4.4                scran_1.16.0               
##  [7] scater_1.16.2               ggplot2_3.3.2              
##  [9] scRNAseq_2.2.0              SingleCellExperiment_1.10.1
## [11] SummarizedExperiment_1.18.2 DelayedArray_0.14.1        
## [13] matrixStats_0.56.0          Biobase_2.48.0             
## [15] GenomicRanges_1.40.0        GenomeInfoDb_1.24.2        
## [17] IRanges_2.22.2              S4Vectors_0.26.1           
## [19] BiocGenerics_0.34.0         AnVIL_1.1.14               
## [21] dplyr_1.0.0                
## 
## loaded via a namespace (and not attached):
##   [1] ggbeeswarm_0.6.0              colorspace_1.4-1             
##   [3] ellipsis_0.3.1                futile.logger_1.4.3          
##   [5] XVector_0.28.0                BiocNeighbors_1.6.0          
##   [7] rstudioapi_0.11               bit64_0.9-7.1                
##   [9] fansi_0.4.1                   RSpectra_0.16-0              
##  [11] interactiveDisplayBase_1.26.3 AnnotationDbi_1.50.1         
##  [13] codetools_0.2-16              knitr_1.29                   
##  [15] jsonlite_1.7.0                Rsamtools_2.4.0              
##  [17] uwot_0.1.8                    shiny_1.5.0                  
##  [19] HDF5Array_1.16.1              BiocManager_1.30.10          
##  [21] compiler_4.0.2                httr_1.4.1                   
##  [23] dqrng_0.2.1                   assertthat_0.2.1             
##  [25] Matrix_1.2-18                 fastmap_1.0.1                
##  [27] cli_2.0.2                     limma_3.44.3                 
##  [29] later_1.1.0.1                 BiocSingular_1.4.0           
##  [31] formatR_1.7                   htmltools_0.5.0              
##  [33] tools_4.0.2                   rsvd_1.0.3                   
##  [35] igraph_1.2.5                  gtable_0.3.0                 
##  [37] glue_1.4.1                    GenomeInfoDbData_1.2.3       
##  [39] rappdirs_0.3.1                Rcpp_1.0.5                   
##  [41] rapiclient_0.1.3              vctrs_0.3.2                  
##  [43] Biostrings_2.56.0             ExperimentHub_1.14.0         
##  [45] DelayedMatrixStats_1.10.1     xfun_0.15                    
##  [47] stringr_1.4.0                 mime_0.9                     
##  [49] lifecycle_0.2.0               irlba_2.3.3                  
##  [51] statmod_1.4.34                XML_3.99-0.4                 
##  [53] AnnotationHub_2.20.0          edgeR_3.30.3                 
##  [55] zlibbioc_1.34.0               scales_1.1.1                 
##  [57] promises_1.1.1                lambda.r_1.2.4               
##  [59] yaml_2.2.1                    curl_4.3                     
##  [61] memoise_1.1.0                 gridExtra_2.3                
##  [63] stringi_1.4.6                 RSQLite_2.2.0                
##  [65] BiocVersion_3.11.1            BiocParallel_1.22.0          
##  [67] rlang_0.4.7                   pkgconfig_2.0.3              
##  [69] bitops_1.0-6                  evaluate_0.14                
##  [71] lattice_0.20-41               purrr_0.3.4                  
##  [73] Rhdf5lib_1.10.1               GenomicAlignments_1.24.0     
##  [75] bit_1.1-15.2                  tidyselect_1.1.0             
##  [77] magrittr_1.5                  bookdown_0.20                
##  [79] R6_2.4.1                      generics_0.0.2               
##  [81] DBI_1.1.0                     pillar_1.4.6                 
##  [83] withr_2.2.0                   RCurl_1.98-1.2               
##  [85] tibble_3.0.3                  crayon_1.3.4                 
##  [87] futile.options_1.0.1          utf8_1.1.4                   
##  [89] rmarkdown_2.3                 viridis_0.5.1                
##  [91] locfit_1.5-9.4                grid_4.0.2                   
##  [93] blob_1.2.1                    digest_0.6.25                
##  [95] xtable_1.8-4                  tidyr_1.1.0                  
##  [97] httpuv_1.5.4                  munsell_0.5.0                
##  [99] beeswarm_0.2.3                viridisLite_0.3.0            
## [101] vipor_0.4.5</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="working-with-large-data.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="formal-work-flows-human-cell-atlas-optimus-pipeline.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
}
});
});
</script>

</body>

</html>
