<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Using an existing workspace | R / Bioconductor in the AnVIL Cloud</title>
  <meta name="description" content="This book contains material for a 2 1/2 hour course offered at the Bioinformatics Community Conference 2020." />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Using an existing workspace | R / Bioconductor in the AnVIL Cloud" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This book contains material for a 2 1/2 hour course offered at the Bioinformatics Community Conference 2020." />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Using an existing workspace | R / Bioconductor in the AnVIL Cloud" />
  
  <meta name="twitter:description" content="This book contains material for a 2 1/2 hour course offered at the Bioinformatics Community Conference 2020." />
  

<meta name="author" content="Martin Morgan" />
<meta name="author" content="Nitesh Turaga" />
<meta name="author" content="Lori Shepherd" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="the-anvil-workspaces-runtimes-buckets-workflows.html"/>
<link rel="next" href="working-with-large-data.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">R / Bioconductor in the AnVIL Cloud</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a><ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#abstract"><i class="fa fa-check"></i>Abstract</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i>Prerequisites</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#learning-objectives"><i class="fa fa-check"></i>Learning objectives</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1</b> Prerequisites</a><ul>
<li class="chapter" data-level="1.1" data-path="prerequisites.html"><a href="prerequisites.html"><i class="fa fa-check"></i><b>1.1</b> Create a google ‘billing account’</a></li>
<li class="chapter" data-level="1.2" data-path="prerequisites.html"><a href="prerequisites.html#create-an-anvil-account"><i class="fa fa-check"></i><b>1.2</b> Create an AnVIL account</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html"><i class="fa fa-check"></i><b>2</b> The AnVIL – workspaces, runtimes, buckets, workflows</a><ul>
<li class="chapter" data-level="2.1" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html#navigation-icons"><i class="fa fa-check"></i><b>2.1</b> Navigation icons</a></li>
<li class="chapter" data-level="2.2" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html#use-an-existing-workspace"><i class="fa fa-check"></i><b>2.2</b> Use an existing workspace</a></li>
<li class="chapter" data-level="2.3" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html#jupyter-notebook-runtimes"><i class="fa fa-check"></i><b>2.3</b> Jupyter notebook runtimes</a></li>
<li class="chapter" data-level="2.4" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html#google-bucket"><i class="fa fa-check"></i><b>2.4</b> Google bucket</a></li>
<li class="chapter" data-level="2.5" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html#notebooks"><i class="fa fa-check"></i><b>2.5</b> Notebooks</a></li>
<li class="chapter" data-level="2.6" data-path="the-anvil-workspaces-runtimes-buckets-workflows.html"><a href="the-anvil-workspaces-runtimes-buckets-workflows.html#workflows"><i class="fa fa-check"></i><b>2.6</b> Workflows</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html"><i class="fa fa-check"></i><b>3</b> Using an existing workspace</a><ul>
<li class="chapter" data-level="3.1" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#orchestrating-single-cell-analysis"><i class="fa fa-check"></i><b>3.1</b> Orchestrating single cell analysis</a></li>
<li class="chapter" data-level="3.2" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#install-and-load-r-packages"><i class="fa fa-check"></i><b>3.2</b> Install and load <em>R</em> packages</a></li>
<li class="chapter" data-level="3.3" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#example-data-as-a-singlecellexperiment"><i class="fa fa-check"></i><b>3.3</b> Example data as a SingleCellExperiment</a></li>
<li class="chapter" data-level="3.4" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#the-quick-start-workflow"><i class="fa fa-check"></i><b>3.4</b> The ‘Quick start’ workflow</a></li>
<li class="chapter" data-level="3.5" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#information-about-the-packages-used-in-this-session"><i class="fa fa-check"></i><b>3.5</b> Information about the packages used in this session</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="working-with-large-data.html"><a href="working-with-large-data.html"><i class="fa fa-check"></i><b>4</b> Working with large data</a><ul>
<li class="chapter" data-level="4.1" data-path="working-with-large-data.html"><a href="working-with-large-data.html#human-cell-atlas-data-discovery"><i class="fa fa-check"></i><b>4.1</b> Human cell atlas data discovery</a></li>
<li class="chapter" data-level="4.2" data-path="working-with-large-data.html"><a href="working-with-large-data.html#r-bioconductor-libraries"><i class="fa fa-check"></i><b>4.2</b> <em>R / Bioconductor</em> libraries</a></li>
<li class="chapter" data-level="4.3" data-path="working-with-large-data.html"><a href="working-with-large-data.html#data-download-and-import"><i class="fa fa-check"></i><b>4.3</b> Data download and import</a></li>
<li class="chapter" data-level="4.4" data-path="working-with-large-data.html"><a href="working-with-large-data.html#delayedarray-representation-of-large-on-disk-data"><i class="fa fa-check"></i><b>4.4</b> ‘DelayedArray’ representation of large on-disk data</a></li>
<li class="chapter" data-level="4.5" data-path="working-with-large-data.html"><a href="working-with-large-data.html#subseting-and-in-memory-location-for-data-exploration"><i class="fa fa-check"></i><b>4.5</b> Subseting and in-memory location for data exploration</a></li>
<li class="chapter" data-level="4.6" data-path="working-with-large-data.html"><a href="working-with-large-data.html#exploration-of-column-cell-annotations"><i class="fa fa-check"></i><b>4.6</b> Exploration of column (cell) annotations</a></li>
<li class="chapter" data-level="4.7" data-path="working-with-large-data.html"><a href="working-with-large-data.html#invariant-cell-annotations"><i class="fa fa-check"></i><b>4.7</b> Invariant cell annotations</a></li>
<li class="chapter" data-level="4.8" data-path="working-with-large-data.html"><a href="working-with-large-data.html#donor-provenance-and-organ-labels"><i class="fa fa-check"></i><b>4.8</b> Donor provenance and organ labels</a></li>
<li class="chapter" data-level="4.9" data-path="working-with-large-data.html"><a href="working-with-large-data.html#sub-setting-data-based-on-cell-attributes"><i class="fa fa-check"></i><b>4.9</b> Sub-setting data based on cell attributes</a></li>
<li class="chapter" data-level="4.10" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#information-about-the-packages-used-in-this-session"><i class="fa fa-check"></i><b>4.10</b> Information about the packages used in this session</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html"><i class="fa fa-check"></i><b>5</b> The AnVIL R / Bioconductor package</a><ul>
<li class="chapter" data-level="5.1" data-path="working-with-large-data.html"><a href="working-with-large-data.html#r-bioconductor-libraries"><i class="fa fa-check"></i><b>5.1</b> <em>R / Bioconductor</em> libraries</a></li>
<li class="chapter" data-level="5.2" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#the-anvil-workspace"><i class="fa fa-check"></i><b>5.2</b> The <em>AnVIL</em> workspace</a></li>
<li class="chapter" data-level="5.3" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#saving-analysis-products"><i class="fa fa-check"></i><b>5.3</b> Saving analysis products</a><ul>
<li class="chapter" data-level="5.3.1" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#an-abbreviated-work-flow"><i class="fa fa-check"></i><b>5.3.1</b> An abbreviated work flow</a></li>
<li class="chapter" data-level="5.3.2" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#saving-results-on-the-runtime-instance"><i class="fa fa-check"></i><b>5.3.2</b> Saving results on the runtime instance</a></li>
<li class="chapter" data-level="5.3.3" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#backing-up-results-to-the-workspace-bucket"><i class="fa fa-check"></i><b>5.3.3</b> Backing up results to the workspace bucket</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#workspace-data-tables"><i class="fa fa-check"></i><b>5.4</b> Workspace DATA ‘TABLES’</a><ul>
<li class="chapter" data-level="5.4.1" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#metadata-discovery-and-export-from-the-hca"><i class="fa fa-check"></i><b>5.4.1</b> Metadata discovery and export from the HCA</a></li>
<li class="chapter" data-level="5.4.2" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#accessing-tables"><i class="fa fa-check"></i><b>5.4.2</b> Accessing TABLES</a></li>
<li class="chapter" data-level="5.4.3" data-path="the-anvil-r-bioconductor-package.html"><a href="the-anvil-r-bioconductor-package.html#what-do-we-learn-about-the-samples-in-this-experiment"><i class="fa fa-check"></i><b>5.4.3</b> What do we learn about the samples in this experiment?</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="using-an-existing-workspace.html"><a href="using-an-existing-workspace.html#information-about-the-packages-used-in-this-session"><i class="fa fa-check"></i><b>5.5</b> Information about the packages used in this session</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="formal-work-flows-human-cell-atlas-optimus-pipeline.html"><a href="formal-work-flows-human-cell-atlas-optimus-pipeline.html"><i class="fa fa-check"></i><b>6</b> Formal work flows: Human Cell Atlas ‘Optimus’ pipeline</a><ul>
<li class="chapter" data-level="6.1" data-path="formal-work-flows-human-cell-atlas-optimus-pipeline.html"><a href="formal-work-flows-human-cell-atlas-optimus-pipeline.html#the-hca-optimus-single-cell-analysis-pipeline"><i class="fa fa-check"></i><b>6.1</b> The HCA Optimus single cell analysis pipeline</a></li>
<li class="chapter" data-level="6.2" data-path="formal-work-flows-human-cell-atlas-optimus-pipeline.html"><a href="formal-work-flows-human-cell-atlas-optimus-pipeline.html#running-a-work-flow"><i class="fa fa-check"></i><b>6.2</b> Running a work flow</a></li>
<li class="chapter" data-level="6.3" data-path="formal-work-flows-human-cell-atlas-optimus-pipeline.html"><a href="formal-work-flows-human-cell-atlas-optimus-pipeline.html#using-data-in-r"><i class="fa fa-check"></i><b>6.3</b> Using data in <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="next-steps.html"><a href="next-steps.html"><i class="fa fa-check"></i><b>7</b> Next steps</a><ul>
<li class="chapter" data-level="7.1" data-path="next-steps.html"><a href="next-steps.html#shiny-apps"><i class="fa fa-check"></i><b>7.1</b> Shiny apps</a></li>
<li class="chapter" data-level="7.2" data-path="next-steps.html"><a href="next-steps.html#protected-data-access"><i class="fa fa-check"></i><b>7.2</b> Protected data access</a></li>
<li class="chapter" data-level="7.3" data-path="next-steps.html"><a href="next-steps.html#other-clouds"><i class="fa fa-check"></i><b>7.3</b> Other clouds</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">R / Bioconductor in the AnVIL Cloud</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="using-an-existing-workspace" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Using an existing workspace</h1>
<p>In this section we learn about</p>
<ul>
<li>Working with Jupyter notebooks</li>
<li>Installing and loading <em>R / Bioconductor</em> packages</li>
<li>Using the <a href="https://osca.bioconductor.org">OSCA</a> book to perform single-cell data</li>
</ul>
<p>Packages used</p>
<ul>
<li><a href="https://bioconductor.org/packages/scRNAseq">scRNAseq</a> – Pre-defined single cell data sets</li>
<li><a href="https://bioconductor.org/packages/scater">scater</a> – Single-cell quality control and normalization</li>
<li><a href="https://bioconductor.org/packages/scran">scran</a> – Feature selection, dimensionality reduction,
clustering, and visualization</li>
</ul>
<div id="orchestrating-single-cell-analysis" class="section level2">
<h2><span class="header-section-number">3.1</span> Orchestrating single cell analysis</h2>
<p>We tackle the ‘quick start’ section of the book ‘Orchestrating single
cell analysis in R / Bioconductor’ (<a href="https://osca.bioconductor.org">OSCA</a>).</p>
<p>Visit <a href="https://osca.bioconductor.org/overview.html#quick-start">Quick start</a>, section 5.5, of OSCA.</p>
</div>
<div id="install-and-load-r-packages" class="section level2">
<h2><span class="header-section-number">3.2</span> Install and load <em>R</em> packages</h2>
<p>We will use several <em>R / Bioconductor</em> packages and their
dependencies. Make sure that the packages are installed in <em>R</em> by
running the following command. Package installation can take quite a
long time; this is an area that is being actively worked on.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co">## packages &lt;- c(&quot;scRNAseq&quot;, &quot;scater&quot;, &quot;scran&quot;, &quot;uwot&quot;)</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="co">## need &lt;- packages[!packages %in% rownames(installed.packages())]</span></a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="co">## BiocManager::install(need, update = FALSE)</span></a></code></pre></div>
<p>Detailed information on <em>R / Bioconductor</em> packages can be found on
the package ‘landing pages’, e.g., for <a href="https://bioconductor.org/packages/scater">scater</a>. Be sure to check
out the vignettes available on the landing page.</p>
<p>Attach the packages we will use into the current <em>R</em>
session. Attaching a package adds the functions and other objects
defined in the package to the <em>R</em> ‘search’ path, so that the commands
can be evaluated.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">suppressPackageStartupMessages</span>({</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">    <span class="kw">library</span>(scRNAseq)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">    <span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb2-4" data-line-number="4">    <span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">})</a></code></pre></div>
</div>
<div id="example-data-as-a-singlecellexperiment" class="section level2">
<h2><span class="header-section-number">3.3</span> Example data as a SingleCellExperiment</h2>
<p>We’ll work with a pre-defined <code>SingleCellExperiment</code> data set; see the
OSCA book for steps to create your own <code>SingleCellExperiment</code>. Check
out the help page <code>?MacoskoRetinaData</code> for more information on this
data set.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">sce &lt;-<span class="st"> </span><span class="kw">MacoskoRetinaData</span>()</a></code></pre></div>
<pre><code>## snapshotDate(): 2020-04-27</code></pre>
<pre><code>## see ?scRNAseq and browseVignettes(&#39;scRNAseq&#39;) for documentation</code></pre>
<pre><code>## loading from cache</code></pre>
<pre><code>## see ?scRNAseq and browseVignettes(&#39;scRNAseq&#39;) for documentation</code></pre>
<pre><code>## loading from cache</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">sce</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 24658 49300 
## metadata(0):
## assays(1): counts
## rownames(24658): KITL TMTC3 ... 1110059M19RIK GM20861
## rowData names(0):
## colnames(49300): r1_GGCCGCAGTCCG r1_CTTGTGCGGGAA ... p1_TAACGCGCTCCT
##   p1_ATTCTTGTTCTT
## colData names(2): cell.id cluster
## reducedDimNames(0):
## altExpNames(0):</code></pre>
<p>Printing <code>sce</code> shows <code>dim: 24658 49300</code>, indicating that there are
24658 rows (genes) and 49300 columns (cells). <code>colData names(2): cell.id cluster</code> indicates that each column (cell) has additional
information on cell id and cluster, perhaps from a previous analysis.</p>
<p>The count data (number of reads from each cell mapping to each gene)
can be extracted from <code>sce</code> using <code>assay()</code>; this is a a 24658 x 49300
matrix, so we subset it to show only the first 10 rows and 3
columns. Many of the counts in typical single cell experiments are
zeros, so the matrix is represented using a sparse matrix
representation defined in the base R Matrix package.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw">assay</span>(sce, <span class="st">&quot;counts&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a></code></pre></div>
<pre><code>## 10 x 3 sparse Matrix of class &quot;dgCMatrix&quot;
##               r1_GGCCGCAGTCCG r1_CTTGTGCGGGAA r1_GCGCAACTGCTC
## KITL                        .               .               1
## TMTC3                       3               .               .
## CEP290                      1               3               .
## 4930430F08RIK               2               1               2
## 1700017N19RIK               .               .               .
## MGAT4C                      .               .               4
## RASSF9                      .               .               .
## LRRIQ1                      .               .               .
## ADGB                        .               .               .
## SLC6A15                     4               1               3</code></pre>
</div>
<div id="the-quick-start-workflow" class="section level2">
<h2><span class="header-section-number">3.4</span> The ‘Quick start’ workflow</h2>
<p>Use the functions <code>perCellQCMetrics()</code> and <code>quickPerCellQC()</code> to
identify cells failing to satisfy quality control metrics. Details of
these functions can be found in the <a href="https://osca.bioconductor.org">OSCA</a> book, on the help
pages <code>?perCellQCMetrics</code>, and in package vignettes.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">is.mito &lt;-<span class="st"> </span><span class="kw">grepl</span>(<span class="st">&quot;^MT-&quot;</span>, <span class="kw">rownames</span>(sce))</a>
<a class="sourceLine" id="cb13-2" data-line-number="2">qcstats &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(sce, <span class="dt">subsets=</span><span class="kw">list</span>(<span class="dt">Mito=</span>is.mito))</a>
<a class="sourceLine" id="cb13-3" data-line-number="3">filtered &lt;-<span class="st"> </span><span class="kw">quickPerCellQC</span>(qcstats, <span class="dt">percent_subsets=</span><span class="st">&quot;subsets_Mito_percent&quot;</span>)</a>
<a class="sourceLine" id="cb13-4" data-line-number="4">sce &lt;-<span class="st"> </span>sce[, <span class="op">!</span>filtered<span class="op">$</span>discard]</a></code></pre></div>
<p>The number of cells has been reduced from 49300 to 45877.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw">dim</span>(sce)</a></code></pre></div>
<pre><code>## [1] 24658 45877</code></pre>
<p>Normalization transforms counts to accommodate differences in ‘library
size’ (total number of reads assayed in each cell) and the
distribution of the counts. There are several approaches to
normalization; we adopt the approach implemented by the
<code>logNormCounts()</code> function. This updates <code>sce</code> to include a second
assay, called <code>&quot;logcounts&quot;</code>.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">sce &lt;-<span class="st"> </span><span class="kw">logNormCounts</span>(sce)</a>
<a class="sourceLine" id="cb16-2" data-line-number="2"><span class="kw">assay</span>(sce, <span class="st">&quot;logcounts&quot;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">3</span>]</a></code></pre></div>
<pre><code>## 10 x 3 sparse Matrix of class &quot;dgCMatrix&quot;
##               r1_GGCCGCAGTCCG r1_CTTGTGCGGGAA r1_GCGCAACTGCTC
## KITL               .               .                0.0670073
## TMTC3              0.14677618      .                .        
## CEP290             0.05060284      0.17027364       .        
## 4930430F08RIK      0.09949075      0.05901921       0.1310400
## 1700017N19RIK      .               .                .        
## MGAT4C             .               .                0.2511625
## RASSF9             .               .                .        
## LRRIQ1             .               .                .        
## ADGB               .               .                .        
## SLC6A15            0.19256085      0.05901921       0.1923511</code></pre>
<p>Feature selection identifies genes that are particularly informative
about between-cell variation in gene expression. The functions
<code>modelGeneVar()</code> and <code>getTopHVGs()</code> identify 9two14 particularly
informative features.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">dec &lt;-<span class="st"> </span><span class="kw">modelGeneVar</span>(sce)</a>
<a class="sourceLine" id="cb18-2" data-line-number="2">hvg &lt;-<span class="st"> </span><span class="kw">getTopHVGs</span>(dec, <span class="dt">prop=</span><span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb18-3" data-line-number="3"><span class="kw">head</span>(hvg)</a></code></pre></div>
<pre><code>## [1] &quot;RHO&quot;     &quot;CALM1&quot;   &quot;MEG3&quot;    &quot;GNGT1&quot;   &quot;SAG&quot;     &quot;RPGRIP1&quot;</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw">length</span>(hvg)</a></code></pre></div>
<pre><code>## [1] 914</code></pre>
<p>Each column of the assay matrix can be though of as a vector defining
the location of a cell in a high-dimensional space. Dimensionality
reduction projects this high dimensional space onto lower dimensions
to enhance visualization. The following commands perform ‘PCA’ and
‘UMAP’ dimensionality reduction. The <code>sce</code> object is updated to record
these reductions, so the calculation can be performed only once. It
takes 2 or 3 minutes to evaluate the following cell.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">1234</span>)</a>
<a class="sourceLine" id="cb22-2" data-line-number="2"><span class="kw">system.time</span>({</a>
<a class="sourceLine" id="cb22-3" data-line-number="3">    sce &lt;-<span class="st"> </span><span class="kw">runPCA</span>(sce, <span class="dt">ncomponents=</span><span class="dv">25</span>, <span class="dt">subset_row=</span>hvg)</a>
<a class="sourceLine" id="cb22-4" data-line-number="4">    sce &lt;-<span class="st"> </span><span class="kw">runUMAP</span>(sce, <span class="dt">dimred =</span> <span class="st">&#39;PCA&#39;</span>, <span class="dt">external_neighbors=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb22-5" data-line-number="5">})</a></code></pre></div>
<pre><code>##    user  system elapsed 
## 152.906   4.526 157.439</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="kw">reducedDimNames</span>(sce)</a></code></pre></div>
<pre><code>## [1] &quot;PCA&quot;  &quot;UMAP&quot;</code></pre>
<p>PCA and UMP can be used to visualize cells in two or three
dimensions. The cells form visual clusters, and <code>cluster_louvain()</code> is
a method to identify cells that are close to one another in the
reduced space.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1">g &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(sce, <span class="dt">use.dimred =</span> <span class="st">&#39;PCA&#39;</span>)</a>
<a class="sourceLine" id="cb26-2" data-line-number="2"><span class="kw">colLabels</span>(sce) &lt;-<span class="st"> </span><span class="kw">factor</span>(igraph<span class="op">::</span><span class="kw">cluster_louvain</span>(g)<span class="op">$</span>membership)</a></code></pre></div>
<p>As a final step in this quick start, cells can be visualized in
reduced dimensional space, colored according to the identified
clusters.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw">plotUMAP</span>(sce, <span class="dt">colour_by=</span><span class="st">&quot;label&quot;</span>)</a></code></pre></div>
<p><img src="BCC2020_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
</div>
<div id="information-about-the-packages-used-in-this-session" class="section level2">
<h2><span class="header-section-number">3.5</span> Information about the packages used in this session</h2>
<p>The <em>R</em> command <code>sessionInfo()</code> captures information about the
versions of software used in the current session. This can be valuable
for performing reproducible analysis.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="kw">sessionInfo</span>()</a></code></pre></div>
<pre><code>## R version 4.0.2 Patched (2020-06-24 r78747)
## Platform: x86_64-apple-darwin17.7.0 (64-bit)
## Running under: macOS High Sierra 10.13.6
## 
## Matrix products: default
## BLAS:   /Users/ma38727/bin/R-4-0-branch/lib/libRblas.dylib
## LAPACK: /Users/ma38727/bin/R-4-0-branch/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] scran_1.16.0                scater_1.16.2              
##  [3] ggplot2_3.3.2               scRNAseq_2.2.0             
##  [5] SingleCellExperiment_1.10.1 SummarizedExperiment_1.18.2
##  [7] DelayedArray_0.14.1         matrixStats_0.56.0         
##  [9] Biobase_2.48.0              GenomicRanges_1.40.0       
## [11] GenomeInfoDb_1.24.2         IRanges_2.22.2             
## [13] S4Vectors_0.26.1            BiocGenerics_0.34.0        
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                  bit64_0.9-7.1                
##  [3] httr_1.4.1                    tools_4.0.2                  
##  [5] R6_2.4.1                      irlba_2.3.3                  
##  [7] vipor_0.4.5                   DBI_1.1.0                    
##  [9] colorspace_1.4-1              withr_2.2.0                  
## [11] tidyselect_1.1.0              gridExtra_2.3                
## [13] bit_1.1-15.2                  curl_4.3                     
## [15] compiler_4.0.2                BiocNeighbors_1.6.0          
## [17] labeling_0.3                  bookdown_0.20                
## [19] scales_1.1.1                  rappdirs_0.3.1               
## [21] stringr_1.4.0                 digest_0.6.25                
## [23] rmarkdown_2.3                 XVector_0.28.0               
## [25] pkgconfig_2.0.3               htmltools_0.5.0              
## [27] dbplyr_1.4.4                  fastmap_1.0.1                
## [29] limma_3.44.3                  rlang_0.4.7                  
## [31] rstudioapi_0.11               RSQLite_2.2.0                
## [33] shiny_1.5.0                   DelayedMatrixStats_1.10.1    
## [35] farver_2.0.3                  generics_0.0.2               
## [37] BiocParallel_1.22.0           dplyr_1.0.0                  
## [39] RCurl_1.98-1.2                magrittr_1.5                 
## [41] BiocSingular_1.4.0            GenomeInfoDbData_1.2.3       
## [43] Matrix_1.2-18                 Rcpp_1.0.5                   
## [45] ggbeeswarm_0.6.0              munsell_0.5.0                
## [47] viridis_0.5.1                 lifecycle_0.2.0              
## [49] stringi_1.4.6                 yaml_2.2.1                   
## [51] edgeR_3.30.3                  zlibbioc_1.34.0              
## [53] BiocFileCache_1.12.0          AnnotationHub_2.20.0         
## [55] grid_4.0.2                    blob_1.2.1                   
## [57] promises_1.1.1                dqrng_0.2.1                  
## [59] ExperimentHub_1.14.0          crayon_1.3.4                 
## [61] lattice_0.20-41               cowplot_1.0.0                
## [63] locfit_1.5-9.4                knitr_1.29                   
## [65] pillar_1.4.6                  igraph_1.2.5                 
## [67] glue_1.4.1                    BiocVersion_3.11.1           
## [69] evaluate_0.14                 BiocManager_1.30.10          
## [71] vctrs_0.3.2                   httpuv_1.5.4                 
## [73] gtable_0.3.0                  purrr_0.3.4                  
## [75] assertthat_0.2.1              xfun_0.15                    
## [77] rsvd_1.0.3                    mime_0.9                     
## [79] xtable_1.8-4                  later_1.1.0.1                
## [81] viridisLite_0.3.0             tibble_3.0.3                 
## [83] AnnotationDbi_1.50.1          beeswarm_0.2.3               
## [85] memoise_1.1.0                 statmod_1.4.34               
## [87] ellipsis_0.3.1                interactiveDisplayBase_1.26.3</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="the-anvil-workspaces-runtimes-buckets-workflows.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="working-with-large-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section"
}
});
});
</script>

</body>

</html>
